arch/x86: drivers/x86/apic_timer: new local APIC timer driver for x86

gnuless opened the pull request:
The existing local APIC timer driver ~had~ has fallen into disuse in favor of the HPET timer by @andyross. As a result it's not been maintained through the last few timer API changes, especially tickless operation. For better or worse, the local APIC timer is making a raging comeback and is required for real-time operation with some hypervisors like ACRN.

Fixes: #16720

Commit:
drivers/timer/apic_timer.c: new local APIC timer for TICKLESS_KERNEL

The existing local APIC timer driver (loapic_timer.c) has bitrotted
and doesn't support TICKLESS_KERNEL, which is the preferred mode of
operation. This patch introduces a completely new driver, called
the APIC timer driver - the name is changed to allow the drivers to
continue to coexist in the short term, and also because "APIC timer"
isn't ambiguous (the I/O APICs do not have timers).

This driver makes no attempt to work with the MVIC timer as the
previous version did, because MVIC support is deprecated.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

Commit:
boards/x86/up_squared: default to new local APIC timer

This is the "flagship" platform for the new local APIC timer driver.
The opportunity is taken clean up the configuration as well, so the
choice of local APIC vs HPET timer requires changing only one Kconfig.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

andyross commented on drivers/timer/apic_timer.c:
```diff
@@ -0,0 +1,378 @@
+
+	k_spinlock_key_t key = k_spin_lock(&lock);
+	ccr = x86_read_loapic(LOAPIC_TIMER_CCR);
+	icr = x86_read_loapic(LOAPIC_TIMER_ICR);
```
Sort of unfortunate naming.  The "ICR" register you'll find in the SDM is the "Interrupt Control Register".  This thing is the LVT Timer Initial Count, which doesn't seem to have a standard acronym.

On the subject of naming, and unrelated to this patch: we need to kill "loapic" dead.  That's a Zephyrism that no one else  understands.  The Intel term is just "APIC", or "Local APIC" if you absolutely have to disambiguate (but you never do, becuase the IO-APIC does a totally different job).  Linux uses "lapic", so I could see that.  But this name is weird.

andyross replied:
More meaty criticism: would be better to cache ICR rather than read it back from the hardware every time.   You're locked anyway, so it can only be a win.  That's what other drivers do with e.g. "last_load".

gnuless replied:
That's an excellent point.  I will do that.

gnuless replied:
And for the record, I didn't make any of `LOAPIC` or `ICR` here, these were pre-existing.  I agree that they're not the best choices.  I'll make a mental note to fix them after we retire the legacy local APIC timer driver.

andyross commented on drivers/timer/apic_timer.c:
```diff
@@ -0,0 +1,378 @@
+	IRQ_CONNECT(CONFIG_APIC_TIMER_IRQ,
+		CONFIG_APIC_TIMER_IRQ_PRIORITY,
+		isr, 0, 0);
+	irq_enable(CONFIG_APIC_TIMER_IRQ);
```
Put the enable at the end.  You don't know what's in the ICR yet so can't predict when the next interrupt will arrive.

gnuless replied:
_Ackshually_.... the ordering here was deliberate.  I seem to recall that writing the initial count before enabling the interrupt in the LVT would result in some buggy APICs never starting the countdown... but I can't find the reference.

I initially thought an early interrupt wouldn't be a problem, but it would be (for starters, the `total_cycles` + ICR invariant is broken, and that's just the beginning).  I'll make the change, since my memory of buggy APICs might just be wildly out-of-date or my brain malfunctioning due to age..

gnuless replied:
On second thought, I'll just stop the timer before initializing it and the leave the ordering as-is. This way both problems (the early interrupt and the possible buggy APICs) are addressed.

gnuless replied:
On third thought, you're right. I'll just move the enable. In the case of sloppy firmware/bootware that interrupt could be queued on the local APIC when we enable it, and that need not have anything at all to do with the timer...

andyross commented on drivers/timer/apic_timer.c:
```diff
@@ -0,0 +1,378 @@
+/*
+ * Based on my understanding of the contract described in
+ * include/drivers/system_timer.h, this function acts as follows:
+ *
```
Better to update the contract in include/drivers/system_timer.h then than repeat it here, no? :)

gnuless replied:
But that's _your_ documentation. :)

andyross commented on drivers/timer/apic_timer.c:
```diff
@@ -0,0 +1,378 @@
+ *
+ * 2. (a) n == K_FOREVER or
+ *    (b) n > MAX_TICKS (which includes n == INT_MAX)
+ *
```
MAX_TICKS in the other drivers is a precision issue.  If the hardware can't handle the requested timeout it waits for as long as it can.  It's not part of the API per se.  The kernel is prepared to take an announce call at any time (as long as it doesn't lie, of course).

gnuless replied:
Right, right. `MAX_TICKS` is my own maximum, so I suppose it's confusing that I am documenting here as part of the API...

andyross commented on drivers/timer/apic_timer.c:
```diff
@@ -0,0 +1,378 @@
+	ticks = total_cycles - last_announcement;
+	ticks += icr - ccr;
+	ticks /= CYCLES_PER_TICK;
+	k_spin_unlock(&lock, key);
```
Move the last two lines out of the lock, they depend only on local state.

gnuless replied:
Will do.

andyross commented on drivers/timer/apic_timer.c:
```diff
@@ -0,0 +1,378 @@
+
+	k_spinlock_key_t key = k_spin_lock(&lock);
+	ccr = x86_read_loapic(LOAPIC_TIMER_CCR);
+	icr = x86_read_loapic(LOAPIC_TIMER_ICR);
```
This pattern appears a lot.  Probably worth defining something like:

```c
u32_t apic_cyc_elapsed()
{
    return x86_read_loapic(LOAPIC_TIMER_CCR) - x86_read_loapic(LOAPIC_TIMER_ICR);
}
```

andyross commented on drivers/timer/apic_timer.c:
```diff
@@ -0,0 +1,378 @@
+	 * If we get here and the CCR isn't zero, then this interrupt is
+	 * stale: it was queued while z_clock_set_timeout() was setting
+	 * a new counter. Just ignore it. See above for more info.
+	 */
```
Trying to think of a way this could mess you up... can't, but still sounds scary.

gnuless replied:
It's bullet-proof :)

gnuless replied:
Also, one word: PendSV

andyross commented on drivers/timer/apic_timer.c:
```diff
@@ -0,0 +1,378 @@
+	icr = x86_read_loapic(LOAPIC_TIMER_ICR);
+	ccr = x86_read_loapic(LOAPIC_TIMER_CCR);
+	total_cycles += (icr - ccr);
+	partial_cycles = CYCLES_PER_TICK - (total_cycles % CYCLES_PER_TICK);
```
FWIW: the complexity involved in calculating "partial cycles" goes away entirely if you use an Invariant TSC as your time base (along the potential loss of precision when interrupts are late).  Just calculate your "desired TSC expiration" based on the last announcement time, subtract from the current value, and that's your new ICR.

gnuless replied:
The driver is written to be useful on systems without a suitable TSC... I don't know if it's worth starting the special-case spaghetti here, but point taken.  If a usable TSC is assumed, the driver could certainly look a lot different.

gnuless replied:
In fact, on that note, I'll assign myself a github to-do to write a TSC-only driver, then the TSC stuff can come out of here entirely. The idea in this driver is to assume the least common denominator, with a tiny concession to the existence of TSCs where the cost/benefit was high.

andyross reviewed (approved):
Some notes.  Can't see anything wrong.  I will say this is awfully comment-heavy for such a short file. :)

gnuless commented on drivers/timer/apic_timer.c:
```diff
@@ -0,0 +1,378 @@
+ * 'total_cycles') is always an integral multiple of CYCLES_PER_TICK.
+ */
+
+static void isr(void *arg); /* I love forward declarations in the morning */
```
Since it bothers you so, I will re-order the file so this is unnecessary. :)

nashif replied:
and please avoid such unrelated comments :)

gnuless replied:
The forward declaration will go away when I re-order, and the hat-tip to Robert Duvall's iconic movie line will go with it.

gnuless commented:
> Some notes. Can't see anything wrong. I will say this is awfully comment-heavy for such a short file. :)

Well, hopefully they were  _helpful_ comments (if not, I'll remove them). Until recently the existing local APIC driver had comments that were practically copy-and-paste from the data sheets.

gnuless commented:
Force push, took most of the advice from @andyross, fixed issues with warnings due to unused variables when not running tickless.

Commit:
tests/kernel/context: add case for CONFIG_APIC_TIMER

Test needs trivial modification to account for new APIC timer code.
Eventually CONFIG_APIC_TIMER_IRQ, CONFIG_LOAPIC_TIMER_IRQ, etc. will
be consolidated into one CONFIG_TIMER_IRQ to reduce the noise a bit.

Signed-off-by: Charles E. Youse <charles.youse@intel.com>

gnuless commented:
Force push: change ordering of irq_enable() as previously suggested by @andyross and add fix to test to account for new timer type (CONFIG_APIC_TIMER).

pfalcon commented:
This driver is broken in the current master, see https://github.com/zephyrproject-rtos/zephyr/pull/17054#issuecomment-506425340
